File: bestia-client/src/Game/World/SurfacePatch.cs
using System;

namespace BestiaBehemothClient.Game.World
{
  /// <summary>
  /// The visible surface of a square of ground: what the client draws past the range it is sent real chunks.
  /// </summary>
  /// <remarks>
  /// Four planes, one value per sample on a 65x65 grid. Height, standing water, material and canopy density -
  /// between them everything a distant hillside needs.
  ///
  /// <para>
  /// No caves, no overhangs, no ore, no props. A patch is a heightfield, so it cannot express a hole, which is
  /// exactly why it is cheap and exactly why it cannot replace the full-detail ring: you can look at a cave
  /// mouth from a kilometre away, but you cannot walk into one you were only sent a patch of.
  /// </para>
  ///
  /// <para>
  /// It also does not move. A patch comes from the server's heightfield, which no player edit can touch, so it
  /// carries no revision and one held is one that stays correct for the life of the world. That is what makes
  /// <see cref="PatchDiskCache"/> safe where <see cref="ClientChunkStore"/> deliberately is not.
  /// </para>
  /// </remarks>
  public sealed class SurfacePatch
  {
    /// <summary>The dry sentinel. A NaN, so it must be tested with <see cref="HasWaterAt"/> and never with <c>==</c>.</summary>
    public const float NoWater = float.NaN;

    public PatchKey Key { get; }

    /// <summary>Terrain surface elevation per sample, in metres. Defined everywhere, sea floor included.</summary>
    public float[] Height { get; }

    /// <summary>Standing water surface per sample in metres, or <see cref="NoWater"/> where the ground is dry.</summary>
    public float[] Water { get; }

    /// <summary>Block id of the topmost material per sample, indexing the same palette a chunk's voxels do.</summary>
    public byte[] Block { get; }

    /// <summary>Canopy cover per sample as 0..255, for scattering distant vegetation without streaming any.</summary>
    public byte[] Canopy { get; }

    public SurfacePatch(PatchKey key, float[] height, float[] water, byte[] block, byte[] canopy)
    {
      Key = key;
      Height = height;
      Water = water;
      Block = block;
      Canopy = canopy;
    }

    public float HeightAt(int i, int j) => Height[PatchGrid.Index(i, j)];

    public float WaterAt(int i, int j) => Water[PatchGrid.Index(i, j)];

    public bool HasWaterAt(int i, int j) => !float.IsNaN(Water[PatchGrid.Index(i, j)]);

    public byte BlockAt(int i, int j) => Block[PatchGrid.Index(i, j)];

    public byte CanopyAt(int i, int j) => Canopy[PatchGrid.Index(i, j)];

    public override string ToString() => $"SurfacePatch[{Key}]";
  }

  /// <summary>
  /// Reads the wire form of a patch. Mirrors the server's <c>SurfacePatchCodec</c>.
  /// </summary>
  /// <remarks>
  /// Four fixed-width planes after a small header, big-endian. Planes rather than one interleaved stream
  /// because the four values have nothing in common and every fourth byte would break the other three's runs -
  /// the same argument <see cref="RleCodec"/> makes for keeping its two streams apart.
  ///
  /// <para>
  /// Heights are quarter-metres above a floor stored once per patch. A quarter of a metre is a twelfth of a
  /// sample spacing at the finest level, far below what a slope four metres wide can show, and storing them
  /// relative keeps neighbouring values numerically close - which is what the deflate on top is relying on.
  /// </para>
  /// </remarks>
  public static class SurfacePatchCodec
  {
    /// <summary>Named on the wire as <c>SURFACE_PATCH_ENCODING_PLANES_V1</c>, so it moves only when the bytes do.</summary>
    public const int Version = 1;

    private const float HeightStep = 0.25f;

    private const int HeaderBytes = 1 + 1 + 4 + 4 + 4;
    private const int BytesPerSample = 6;

    /// <summary>Exact size of a payload, so a truncated one is caught before it is read rather than after.</summary>
    public static int EncodedSize => HeaderBytes + PatchGrid.SampleCount * BytesPerSample;

    public static SurfacePatch Decode(byte[] bytes)
    {
      if (bytes == null || bytes.Length != EncodedSize)
      {
        throw new FormatException($"a patch payload is {EncodedSize} bytes, got {bytes?.Length ?? 0}");
      }

      var at = 0;
      int version = bytes[at++];
      if (version != Version)
      {
        throw new FormatException($"patch encoded with surface-patch version {version}, this build reads {Version}");
      }

      int level = bytes[at++];
      var x = ReadInt(bytes, ref at);
      var y = ReadInt(bytes, ref at);
      var floorSteps = ReadInt(bytes, ref at);

      var count = PatchGrid.SampleCount;
      var height = new float[count];
      var water = new float[count];
      var block = new byte[count];
      var canopy = new byte[count];

      for (var i = 0; i < count; i++)
      {
        height[i] = Metres(ReadShort(bytes, ref at), floorSteps);
      }

      for (var i = 0; i < count; i++)
      {
        // Zero is the dry marker, which is why heights in this plane are stored from one rather than zero.
        var raw = ReadShort(bytes, ref at);
        water[i] = raw == 0 ? SurfacePatch.NoWater : Metres(raw - 1, floorSteps);
      }

      Array.Copy(bytes, at, block, 0, count);
      at += count;
      Array.Copy(bytes, at, canopy, 0, count);

      return new SurfacePatch(new PatchKey(level, x, y), height, water, block, canopy);
    }

    private static float Metres(int steps, int floorSteps) => (steps + floorSteps) * HeightStep;

    private static int ReadShort(byte[] bytes, ref int at)
    {
      var value = (bytes[at] << 8) | bytes[at + 1];
      at += 2;
      return value;
    }

    private static int ReadInt(byte[] bytes, ref int at)
    {
      var value = (bytes[at] << 24) | (bytes[at + 1] << 16) | (bytes[at + 2] << 8) | bytes[at + 3];
      at += 4;
      return value;
    }
  }
}

File: worldgen/src/main/kotlin/net/bestia/worldgen/lod/SurfacePatch.kt
package net.bestia.worldgen.lod

/**
 * The visible surface of a square of ground, sampled on a regular grid instead of materialised into voxels.
 *
 * What a client draws past the range where it is sent real chunks. Four planes, one value per sample, and
 * between them they are everything a distant hillside needs: where the ground is, where standing water is,
 * what the ground is made of, and how wooded it is.
 *
 * ### What it deliberately does not carry
 *
 * No caves, no overhangs, no ore, no strata, no props. A patch is a *heightfield* - one surface per column -
 * so it cannot express a hole, and it does not try to. That is what makes it cheap, and it is also why the
 * full-detail ring cannot be replaced by it: you can look at a cave mouth from a kilometre away, but you
 * cannot walk into one you were only sent a patch of.
 *
 * It also carries nothing a player could not see. Ore is under the ground and cave interiors are inside it;
 * both are absent here for the same reason the world seed is absent from the wire.
 *
 * ### Player edits are not in it
 *
 * A patch comes from the heightfield, which is immutable, so digging does not move it and it needs no
 * revision. `ChunkService.surfaceElevationAt` already makes the same trade on the movement path. At four
 * metres a sample the largest thing a player can dig is smaller than the grid.
 */
class SurfacePatch(
  val pos: PatchPos,

  /**
   * Terrain surface elevation per sample, in metres. Absolute, like every other elevation in the generator.
   *
   * Defined everywhere, sea floor included - [water] is what says whether you can see the ground or the
   * water over it.
   */
  val height: FloatArray,

  /**
   * Standing water surface per sample in metres, or [NO_WATER] where the ground is dry.
   *
   * Sea and lakes only. River water is the one water body whose surface is not level, so it comes from a
   * per-chunk sampler rather than from the world tier - and at four metres a sample a channel is a sample
   * or two wide, which is a line of dots rather than a river. Distant rivers want the map's own treatment,
   * not this one.
   */
  val water: FloatArray,

  /** `BlockType.id` of the topmost block per sample, which is what decides how it is drawn. */
  val block: ByteArray,

  /** Canopy cover per sample as `0..255`, the same void fraction `LayerId.CANOPY_COVER` holds. */
  val canopy: ByteArray
) {

  init {
    val expected = PatchGrid.SAMPLES * PatchGrid.SAMPLES
    require(height.size == expected && water.size == expected) {
      "a patch holds $expected samples, got ${height.size} heights and ${water.size} waters"
    }
    require(block.size == expected && canopy.size == expected) {
      "a patch holds $expected samples, got ${block.size} blocks and ${canopy.size} canopy values"
    }
  }

  fun heightAt(i: Int, j: Int) = height[index(i, j)]

  fun waterAt(i: Int, j: Int) = water[index(i, j)]

  /** Whether this sample stands under water. Not `!= NO_WATER`: the sentinel is a NaN, so it never equals. */
  fun hasWaterAt(i: Int, j: Int) = !water[index(i, j)].isNaN()

  override fun toString() = "SurfacePatch[$pos]"

  companion object {

    /**
     * The dry sentinel.
     *
     * `NaN` rather than a very negative number, so that a caller who forgets to test it gets an arithmetic
     * result that is obviously wrong everywhere it is used, rather than a waterline at the bottom of the
     * world that looks like a rendering bug. The cost is that it must be tested with [hasWaterAt] and never
     * with `==`, since a NaN does not equal itself.
     */
    const val NO_WATER = Float.NaN

    fun index(i: Int, j: Int) = j * PatchGrid.SAMPLES + i
  }
}
